File: studio/AuraStudio/Sources/AuraStudio/Services/BackgroundTaskCenter.swift
import Foundation

/// PLAN-studio-rendimiento.md Fase 4 punto 1: una única fuente de verdad
/// para "qué está corriendo en segundo plano ahora mismo", que reemplaza
/// los booleanos sueltos de `LibraryViewModel` (`isProcessing`,
/// `isFetchingArtistImages`, `isVerifyingDevice`, `isFetchingVideoPosters`,
/// `isApplyingRecommendedCovers`, más el `@State isEnriching` que vivía
/// en `MediaSectionView`) -- cada uno era su propio interruptor, sin
/// cola, sin cancelación uniforme, y sin un solo lugar donde mostrarlos
/// todos.
///
/// Diagnóstico §0.8: "bien hechos y patrón a copiar" ya señalaba al sync
/// (`DeviceActivityBar`) y a la migración de carátulas ST-141
/// (`CoverNormalizationBar`) como el modelo -- esto generaliza esa idea
/// a cualquier operación larga, no solo a esas dos.
@MainActor
final class BackgroundTaskCenter: ObservableObject {
    /// PLAN-studio-rendimiento-2.md Fase 6 (ST-186): de qué TIPO es una
    /// tarea, para poder decir "de estas, una a la vez".
    ///
    /// No es una categoría decorativa: dos operaciones del mismo tipo
    /// pelean por el mismo recurso. Dos enriquecimientos a la vez
    /// (pedir "Buscar información en línea" desde Álbumes y desde
    /// Canciones, que hoy nada impide) se reparten el mismo cupo de 1
    /// pedido/segundo de MusicBrainz y tardan el doble cada uno, sin que
    /// nada vaya más rápido. Tipos DISTINTOS sí corren en paralelo:
    /// enriquecer mientras se sincroniza es perfectamente razonable.
    enum TaskKind: Sendable, Equatable {
        /// Metadata en línea (MusicBrainz/LRCLIB): cupo compartido.
        case enrichment
        /// Carátulas y pósters en línea (Cover Art Archive, Deezer,
        /// TMDB, fanart.tv).
        case artwork
        /// Trabajo local sobre archivos (importar, editar en lote,
        /// aplicar una carátula ya elegida).
        case files
        /// Lo que no necesita exclusión con nada.
        case other
    }

    enum Progress: Equatable {
        /// "N de M" -- se conoce el total de antemano (enriquecer,
        /// edición en lote, importar).
        case determinate(completed: Int, total: Int)
        /// No se puede estimar cuánto falta (verificar dispositivo,
        /// buscar carátulas recomendadas).
        case indeterminate

        var fraction: Double? {
            guard case let .determinate(completed, total) = self, total > 0 else { return nil }
            return Double(completed) / Double(total)
        }
    }

    /// Una tarea visible en el centro. `ObservableObject` propio (no
    /// solo un struct) para que actualizar SU progreso no publique
    /// `objectWillChange` de todo `BackgroundTaskCenter` -- una fila del
    /// popover se repinta sola, no la lista entera ni la ventana.
    @MainActor
    final class TaskHandle: ObservableObject, Identifiable {
        let id = UUID()
        /// En español, de cara al usuario -- "Buscando información en
        /// línea…", "Importando 200 archivos…".
        let title: String
        let kind: TaskKind
        @Published private(set) var progress: Progress
        /// Detalle opcional bajo el título -- "Álbum 12 de 40".
        @Published var statusText: String?
        /// `nil` mientras corre bien. Si se pone, la fila del popover lo
        /// muestra y dejar de existir depende de `finish`/`cancel`, igual
        /// que una que terminó bien -- un error no dura para siempre.
        @Published private(set) var errorText: String?
        private(set) var isCancelled = false
        private let onCancelRequested: (@MainActor () -> Void)?

        /// `true` cuando la tarea no ofrece cancelar (encargo del
        /// dueño: algunas operaciones, como escribir el catálogo, no se
        /// pueden interrumpir a medias sin arriesgar corromper algo).
        var isCancellable: Bool { onCancelRequested != nil }

        init(title: String, kind: TaskKind = .other, progress: Progress,
             onCancelRequested: (@MainActor () -> Void)? = nil) {
            self.title = title
            self.kind = kind
            self.progress = progress
            self.onCancelRequested = onCancelRequested
        }

        func update(_ progress: Progress, statusText: String? = nil) {
            self.progress = progress
            if let statusText { self.statusText = statusText }
        }

        func fail(_ message: String) {
            errorText = message
        }

        func requestCancel() {
            guard !isCancelled else { return }
            isCancelled = true
            onCancelRequested?()
        }
    }

    @Published private(set) var tasks: [TaskHandle] = []

    var isEmpty: Bool { tasks.isEmpty }
    var count: Int { tasks.count }

    /// Progreso agregado para el anillo de la barra de estado: el
    /// promedio de las tareas determinadas: las indeterminadas no
    /// aportan una fracción, así que no cuentan para el promedio, pero
    /// SU sola presencia ya alcanza para que `count > 0` muestre el
    /// indicador (un anillo animado sin fracción fija, como el de
    /// Finder mientras calcula "Preparando…").
    var aggregateFraction: Double? {
        let fractions = tasks.compactMap(\.progress.fraction)
        guard !fractions.isEmpty else { return nil }
        return fractions.reduce(0, +) / Double(fractions.count)
    }

    /// Registra una tarea nueva y la deja visible de inmediato. El
    /// llamador se queda con el `TaskHandle` para actualizar su
    /// progreso y llamar `finish(_:)` cuando termine -- `defer { center.
    /// finish(handle) }` es el patrón esperado, para que un `throw` a
    /// mitad de camino no deje la tarea pegada en el centro para
    /// siempre.
    @discardableResult
    func begin(title: String, kind: TaskKind = .other, progress: Progress = .indeterminate,
              onCancelRequested: (@MainActor () -> Void)? = nil) -> TaskHandle {
        let handle = TaskHandle(title: title, kind: kind, progress: progress,
                                onCancelRequested: onCancelRequested)
        tasks.append(handle)
        return handle
    }

    /// ¿Hay ya una tarea de este tipo corriendo? (ST-186.) Quien va a
    /// empezar una que pelea por un recurso compartido pregunta acá en
    /// vez de llevar su propio booleano suelto -- que es lo que ST-156
    /// dejó anotado como pendiente: cada operación tenía su interruptor,
    /// sin un solo lugar donde mirarlos todos.
    func isRunning(_ kind: TaskKind) -> Bool {
        tasks.contains { $0.kind == kind && !$0.isCancelled }
    }

    func finish(_ handle: TaskHandle) {
        tasks.removeAll { $0.id == handle.id }
    }
}

File: studio/windows/AuraStudio.App/Services/BackgroundTaskCenter.cs
using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;
using AuraStudio.Core.Library;

namespace AuraStudio.App.Services;

/// <summary>
/// Una tarea de fondo visible para el usuario (ST-203; paridad con
/// <c>BackgroundTaskCenter.TaskHandle</c> de ST-156).
///
/// <para>Es observable <b>ella</b>, y no solo el centro, para que mover su
/// avance repinte su renglón y no la lista entera: una carga de 12 000
/// elementos avisa decenas de veces, y cada aviso no puede costar redibujar
/// todo lo que esté mirando el centro.</para>
/// </summary>
public sealed partial class BackgroundTaskHandle : ObservableObject
{
    private readonly Action? _onCancelRequested;

    internal BackgroundTaskHandle(string title, BackgroundTaskProgress progress, Action? onCancelRequested)
    {
        Title = title;
        Progress = progress;
        _onCancelRequested = onCancelRequested;
    }

    public Guid Id { get; } = Guid.NewGuid();

    /// <summary>En español y de cara al usuario: "Cargando biblioteca…".</summary>
    public string Title { get; }

    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(Fraction))]
    [NotifyPropertyChangedFor(nameof(IsDeterminate))]
    [NotifyPropertyChangedFor(nameof(Detail))]
    public partial BackgroundTaskProgress Progress { get; private set; }

    /// <summary>Detalle bajo el título; si está vacío se usa el "N de M" del avance.</summary>
    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(Detail))]
    public partial string StatusText { get; set; } = "";

    /// <summary>
    /// Vacío mientras va bien. Si se llena, la tarea lo muestra y sigue
    /// existiendo hasta que quien la abrió la cierre: <b>un error no dura para
    /// siempre</b>, pero tampoco desaparece antes de que alguien lo lea.
    /// </summary>
    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(HasError))]
    public partial string ErrorText { get; private set; } = "";

    public bool HasError => ErrorText.Length > 0;

    public bool IsCancelled { get; private set; }

    /// <summary>
    /// <c>false</c> cuando la tarea no ofrece cancelar. Hay operaciones que no
    /// se pueden interrumpir a medias sin arriesgar dejar algo corrupto —
    /// escribir el catálogo, por ejemplo—, y ofrecer un botón que no se puede
    /// cumplir es peor que no ofrecerlo.
    /// </summary>
    public bool IsCancellable => _onCancelRequested is not null;

    public double? Fraction => Progress.Fraction;

    public bool IsDeterminate => Progress.IsDeterminate;

    /// <summary>Lo que se lee bajo el título: el detalle propio, o el "N de M".</summary>
    public string Detail => StatusText.Length > 0 ? StatusText : Progress.CountText;

    public void Update(BackgroundTaskProgress progress, string? statusText = null)
    {
        Progress = progress;
        if (statusText is not null) StatusText = statusText;
    }

    public void Fail(string message) => ErrorText = message;

    public void RequestCancel()
    {
        if (IsCancelled) return;

        IsCancelled = true;
        _onCancelRequested?.Invoke();
    }
}

/// <summary>
/// Lo que la app está haciendo por su cuenta, dicho en un solo lugar (ST-203;
/// paridad con <c>BackgroundTaskCenter.swift</c> de ST-156).
///
/// <para>Antes, el trabajo de fondo o no se veía —la biblioteca cargando, los
/// tamaños midiéndose— o se anunciaba pisando <c>StatusMessage</c>, que es el
/// renglón donde se contesta lo que el usuario acaba de pedir. Una tarea que
/// nadie pidió no puede tapar la respuesta a lo que sí se pidió.</para>
///
/// <para><b>Todo acá se toca desde el hilo de interfaz.</b> Quien trabaja en el
/// pool vuelve al despachador para mover el avance, igual que el resto de la
/// app (ST-131).</para>
/// </summary>
public sealed partial class BackgroundTaskCenter : ObservableObject
{
    /// <summary>Las tareas vivas, en el orden en que empezaron.</summary>
    public ObservableCollection<BackgroundTaskHandle> Tasks { get; } = [];

    public bool IsEmpty => Tasks.Count == 0;

    public int Count => Tasks.Count;

    /// <summary>
    /// El avance de todas juntas, para un solo indicador. Las indeterminadas no
    /// promedian —no tienen fracción que aportar—, pero su presencia ya basta
    /// para que el indicador se vea.
    /// </summary>
    public double? AggregateFraction =>
        BackgroundTaskProgress.Aggregate(Tasks.Select(task => task.Progress));

    /// <summary>
    /// La que se muestra cuando hay lugar para una sola: la primera. Con una
    /// tarea es la que hay; con varias, la más vieja — que es la que el usuario
    /// lleva más rato esperando.
    /// </summary>
    public BackgroundTaskHandle? Current => Tasks.Count > 0 ? Tasks[0] : null;

    /// <summary>
    /// Registra una tarea y la deja visible ya. Quien llama se queda con lo
    /// devuelto para mover el avance y <b>tiene que llamar a
    /// <see cref="Finish"/></b> cuando termine — con <c>try/finally</c>, para
    /// que un error a mitad no deje la tarea pegada para siempre.
    /// </summary>
    public BackgroundTaskHandle Begin(
        string title, BackgroundTaskProgress progress = default, Action? onCancelRequested = null)
    {
        var handle = new BackgroundTaskHandle(title, progress, onCancelRequested);
        handle.PropertyChanged += OnTaskChanged;
        Tasks.Add(handle);
        Changed();
        return handle;
    }

    public void Finish(BackgroundTaskHandle handle)
    {
        handle.PropertyChanged -= OnTaskChanged;
        if (Tasks.Remove(handle)) Changed();
    }

    private void OnTaskChanged(object? sender, System.ComponentModel.PropertyChangedEventArgs e)
    {
        if (e.PropertyName == nameof(BackgroundTaskHandle.Progress)) OnPropertyChanged(nameof(AggregateFraction));
    }

    private void Changed()
    {
        OnPropertyChanged(nameof(IsEmpty));
        OnPropertyChanged(nameof(Count));
        OnPropertyChanged(nameof(Current));
        OnPropertyChanged(nameof(AggregateFraction));
    }
}
